apis/net/ieee802154/src/rx.rs

```diff
@@ -0,0 +1,310 @@
+        }
+    }
+}
+impl<'buf, const N: usize, S: Syscalls, C: Config> RxOperator
+    for RxSingleBufferOperator<'buf, N, S, C>
+{
+    fn receive_frame(&mut self) -> Result<&mut Frame, ErrorCode> {
+        if self.buf.has_frame() {
+            Ok(self.buf.next_frame())
+        } else {
+            // If no frame is there, wait until one comes, then return it.
+
+            Ieee802154::<S, C>::receive_frame_single_buf(self.buf)?;
+
+            // Safety: kernel schedules an upcall iff a new frame becomes available,
+            // i.e. when it increments `read_index`.
+            Ok(self.buf.next_frame())
+        }
+    }
+}
+
+// Reception
+impl<S: Syscalls, C: Config> Ieee802154<S, C> {
+    fn receive_frame_single_buf<const N: usize>(
+        buf: &mut RxRingBuffer<N>,
+    ) -> Result<(), ErrorCode> {
+        let called: Cell<Option<(u32,)>> = Cell::new(None);
+        share::scope::<
+            (
+                AllowRw<_, DRIVER_NUM, { allow_rw::READ }>,
+                Subscribe<_, DRIVER_NUM, { subscribe::FRAME_RECEIVED }>,
+            ),
+            _,
+            _,
+        >(|handle| {
+            let (allow_rw, subscribe) = handle.split();
+            S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(allow_rw, buf.as_mut_byte_slice())?;
+            S::subscribe::<_, _, C, DRIVER_NUM, { subscribe::FRAME_RECEIVED }>(subscribe, &called)?;
+
+            loop {
+                S::yield_wait();
+                if let Some((_lqi,)) = called.get() {
+                    // At least one frame was received.
+                    return Ok(());
+                }
+            }
+        })
+    }
+}
+
+// This module is to protect RxRingBufferInKernel private API.
+mod alternate_receive_buffers {
+    use super::*;
+
+    pub(super) struct RxRingBufferInKernel<'buf, const N: usize, S: Syscalls, C: Config> {
+        lent_buf: *mut RxRingBuffer<N>,
+        covariance_phantom: PhantomData<&'buf mut RxRingBuffer<N>>,
+        sc: PhantomData<(S, C)>,
+    }
+
+    impl<'buf, const N: usize, S: Syscalls, C: Config> Drop for RxRingBufferInKernel<'buf, N, S, C> {
+        // Unshares the buffer to prevent kernel accessing no longer valid memory.
+        fn drop(&mut self) {
+            share::scope::<AllowRw<_, DRIVER_NUM, { allow_rw::READ }>, _, _>(|handle| {
+                let _ = S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(handle, &mut []);
+            });
+        }
+    }
+
+    impl<'buf, const N: usize, S: Syscalls, C: Config> RxRingBufferInKernel<'buf, N, S, C> {
+        fn new(buf: &mut RxRingBuffer<N>) -> Self {
+            Self {
+                lent_buf: buf as *mut RxRingBuffer<N>,
+                covariance_phantom: PhantomData,
+                sc: PhantomData,
+            }
+        }
+
+        pub(super) fn share_initial(buf: &mut RxRingBuffer<N>) -> Result<Self, ErrorCode> {
+            Self::share_unscoped(buf.as_mut_byte_slice())?;
+            Ok(Self::new(buf))
```

Comment:
> I need help with understanding the Miri check failure. I'm not sure what is exactly believed to be UB and why.

`share_unscoped` is unsound (see further comments below), which allows `share_initial` to violate Rust's memory aliasing rules.

The first line of `share_initial`:
```rust
Self::share_unscoped(buf.as_mut_byte_slice())?;
```
creates a temporary borrow of `buf`, and passes that borrow to `share_unscoped`. Then, on the second line:
```rust
Ok(Self::new(buf))
```
`buf` is used again, which invalidates the temporary borrow of `buf`. This results in two mutable pointers pointing at the same memory: the returned RxRingBufferInKernel and the pointer that was given to the kernel by `share_unscoped`.

Because of the aliasing, when the kernel tries to access the buffer, it causes undefined behavior (this is the UB reported by Miri in the CI failure).

Reply:
What I'm trying to achieve is to split `&mut [u8]` (the mutable reference to the buffer) into:
1. `*mut [u8]` - the mutable pointer to the buffer used in `share_unscoped`, with manual lifetime management;
2. `<PhantomData<&mut [u8]> - a phantom to take advantage of covariance and thus preserve lifetime of the original mutable reference to the buffer.

This way, at `swap`, I want to unsplit those two parts and restore the original mutable reference again.

I have been told that existence of two mutable references pointing to the same memory is unsound, but I don't see how this is the case.